### PR TITLE
BioMed Central Harvester & Normalizer 

### PR DESCRIPTION
**Pull request by caseyrollins**

### CODE REVIEW

File: providers/com/biomedcentral/harvester.py [hunk 1]
```diff
@@ -1,5 +1,47 @@
...
+
+        while total_processed < total:
+            response = self.requests.get(furl(url).add(query_params={
+                'p': 100,
+                's': total_processed
+            }).url)
+
+            records = response.json()['records']
+            for record in records:
+                print(record['identifier'])
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: Remove the prints here.

File: providers/com/biomedcentral/harvester.py [hunk 2]
```diff
@@ -1,5 +1,47 @@
...
+
+        start_date = start_date.date()
+        end_date = end_date.date()
+
+        # API only accepts a specific date, not a date range, for retrieving articles
+        # so we must create our own list of dates
+        dates = [start_date + timedelta(n) for n in range((end_date - start_date).days + 1)]
+
+        for date in dates:
+            return self.fetch_records(furl(self.url).set(query_params={
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: Returning here will only ever run the first date. You could use `yield from` to iterate over all the days

File: providers/com/biomedcentral/harvester.py [hunk 3]
```diff
@@ -1,5 +1,47 @@
...
+                'q': 'date:{}'.format(date),
+            }).url)
+
+    def fetch_records(self, url):
+        resp = self.requests.get(url)
+        total = int(resp.json()['result'][0]['total'])
+        total_processed = 0
+
+        while total_processed < total:
+            response = self.requests.get(furl(url).add(query_params={
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: If you move this below the for loop you'll use all the results that you get back from the first request. As is you'll be making the first request twice.

File: providers/com/biomedcentral/harvester.py [hunk 4]
```diff
@@ -1,5 +1,47 @@
...
+                's': total_processed
+            }).url)
+
+            records = response.json()['records']
+            for record in records:
+                print(record['identifier'])
+                print(record)
+                yield (record['identifier'], record)
+
+            total_processed += 100
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: Prefer to use `+= len(records)` here. If someone decides to change the page size then everything will continue to work.

File: providers/com/biomedcentral/harvester.py [hunk 5]
```diff
@@ -1,5 +1,47 @@
-from share.harvest.harvester import Harvester
+from datetime import timedelta
+from furl import furl
+
+from share import Harvester
+from django.conf import settings
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: use pep8 style imports

```python
import things_from_the_std_lib

import third_party_one.mod
import third_path_one.other_mod

import third_party_two

import things_we_wrote
```

**Comment by caseyrollins** [comment]
@chrisseto 🐘  (changed quite a few things btw)

File: providers/com/biomedcentral/harvester.py [hunk 6]
```diff
@@ -1,5 +1,51 @@
...
+
+        # API only accepts a specific date, not a date range, for retrieving articles
+        # so we must create our own list of dates
+        dates = [start_date + timedelta(n) for n in range((end_date - start_date).days + 1)]
+
+        for date in dates:
+            yield from self.fetch_records(date)
+
+    def fetch_records(self, date):
+        self.offset = 1
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: What's the difference between `offset` and `total_processed`?

File: providers/com/biomedcentral/harvester.py [hunk 7]
```diff
@@ -1,5 +1,51 @@
...
+from django.conf import settings
+
+from share import Harvester
 
 
 class BiomedCentralHarvester(Harvester):
-    pass
+    url = 'http://api.springer.com/meta/v1/json'
+    page_size = 100
+    offset = 1
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: Define instance level variables in the `__init__` method

File: providers/com/biomedcentral/harvester.py (hunk 6 above)
Comment (caseyrollins) [inline providers/com/biomedcentral/harvester.py]: Ah, yeah, they're the same.

**Comment by caseyrollins** [comment]
@chrisseto 🐵

File: providers/com/biomedcentral/harvester.py [hunk 8]
```diff
@@ -1,5 +1,52 @@
...
+from django.conf import settings
+
+from share import Harvester
 
 
 class BiomedCentralHarvester(Harvester):
-    pass
+
+    def __init__(self, app_config):
+        super().__init__('providers.com.biomedcentral.apps.AppConfig')
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: `super().__init__(app_config)`

File: providers/com/biomedcentral/harvester.py [hunk 9]
```diff
@@ -1,5 +1,52 @@
...
+        self.url = 'http://api.springer.com/meta/v1/json'
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: 📐

File: providers/com/biomedcentral/harvester.py [hunk 10]
```diff
@@ -1,5 +1,52 @@
...
+        self.page_size = 100
+        self.offset = 1
+
+    def do_harvest(self, start_date, end_date):
+        if not settings.BIOMEDCENTRAL_API_KEY:
+            raise Exception('BioMed Central api key not provided')
+
+        start_date = start_date.date()
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: 📐

File: providers/com/biomedcentral/normalizer.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    title = ctx.title
+    contributors = ctx.creators['*']
+    description = ctx.abstract
+    published = ctx.publicationDate
+    doi = ctx.doi
+    # subject = ctx.genre
+
+
+class Contributor(Parser):
+    person = ctx
```
Comment (chrisseto) [inline providers/com/biomedcentral/normalizer.py]: add `cited_name` and `order_cited`

File: providers/com/biomedcentral/harvester.py [hunk 12]
```diff
@@ -1,5 +1,52 @@
...
 
 
 class BiomedCentralHarvester(Harvester):
-    pass
+
+    def __init__(self, app_config):
+        super().__init__(app_config)
+        self.offset = 1
+        self.page_size = 100
+        self.url = 'http://api.springer.com/meta/v1/json'
```
Comment (chrisseto) [inline providers/com/biomedcentral/harvester.py]: `https://`

**Pull request merged**